File: ios/PatientInvitation.swift
import Foundation
import OSLog
import Supabase
import SwiftUI

enum InvitationKind: String, Encodable, Sendable {
    case initial
}

/// Response from the `create-patient-invitation` Edge Function.
/// `invitationUrl` is used exactly as returned — the token is never parsed.
nonisolated struct PatientInvitation: Decodable, Sendable {
    let invitationId: UUID
    let invitationUrl: String
    let expiresAt: String
}

private struct CreatePatientInvitationRequest: Encodable {
    let patientId: UUID
    let kind: InvitationKind
}

enum PatientInvitationPreviewStatus: String, Codable, Sendable {
    case valid
    case expired
    case claimed
    case cancelled
    case invalid
}

/// Response from `get-patient-invitation`. Patient-only fields are present
/// when `status` is `valid`.
nonisolated struct PatientInvitationPreview: Decodable, Sendable {
    let status: PatientInvitationPreviewStatus
    let therapistDisplayName: String?
    let expiresAt: String?
}

private struct GetPatientInvitationRequest: Encodable {
    let token: String
}

nonisolated struct ClaimedPatientInvitation: Decodable, Sendable {
    let patientId: UUID
    let therapistId: UUID
}

enum PatientInvitationClaimError: LocalizedError {
    case status(PatientInvitationPreviewStatus)
    case failed

    var errorDescription: String? {
        switch self {
        case .status(.expired): L10n.inviteExpiredTitle
        case .status(.claimed): L10n.inviteClaimedTitle
        case .status(.cancelled): L10n.inviteCancelledTitle
        case .status(.invalid): L10n.inviteInvalidTitle
        case .status(.valid), .failed: L10n.patientActivationClaimFailed
        }
    }
}

/// Creates patient invitations through the shared authenticated Supabase client.
@MainActor
final class PatientInvitationService {
    private let client: SupabaseClient

    init(client: SupabaseClient) {
        self.client = client
    }

    func createPatientInvitation(
        patientId: UUID,
        kind: InvitationKind = .initial
    ) async throws -> PatientInvitation {
        guard SupabaseConfig.isConfigured else { throw AuthError.notConfigured }
        let invitation: PatientInvitation = try await client.functions.invoke(
            "create-patient-invitation",
            options: FunctionInvokeOptions(
                body: CreatePatientInvitationRequest(patientId: patientId, kind: kind)
            )
        )
        AppLog.store.info("Patient invitation created")
        return invitation
    }

    /// Public preview of an invitation token. Callable without signing in.
    /// Does not claim the invitation or create an auth user.
    func getPatientInvitation(token: String) async throws -> PatientInvitationPreview {
        guard SupabaseConfig.isConfigured else { throw AuthError.notConfigured }
        let preview: PatientInvitationPreview = try await client.functions.invoke(
            "get-patient-invitation",
            options: FunctionInvokeOptions(
                body: GetPatientInvitationRequest(token: token)
            )
        )
        AppLog.store.info(
            "Invitation preview status: \(preview.status.rawValue, privacy: .public)"
        )
        return preview
    }

    /// Claims the invitation for the current authenticated (anonymous) user.
    /// Creates `patient_access` only on the server.
    func claimPatientInvitation(token: String) async throws -> ClaimedPatientInvitation {
        guard SupabaseConfig.isConfigured else { throw AuthError.notConfigured }
        do {
            let claimed: ClaimedPatientInvitation = try await client.functions.invoke(
                "claim-patient-invitation",
                options: FunctionInvokeOptions(
                    body: GetPatientInvitationRequest(token: token)
                )
            )
            AppLog.store.info("Patient invitation claimed")
            return claimed
        } catch let FunctionsError.httpError(_, data) {
            if let preview = try? JSONDecoder().decode(PatientInvitationPreview.self, from: data),
               preview.status != .valid {
                throw PatientInvitationClaimError.status(preview.status)
            }
            AppLog.store.error("Patient invitation claim failed")
            throw PatientInvitationClaimError.failed
        } catch let error as PatientInvitationClaimError {
            throw error
        } catch {
            AppLog.store.error(
                "Patient invitation claim failed: \(error.localizedDescription, privacy: .public)"
            )
            throw PatientInvitationClaimError.failed
        }
    }
}

/// Native share sheet. The app has no existing share helper.
struct ActivityShareSheet: UIViewControllerRepresentable {
    let items: [Any]

    func makeUIViewController(context: Context) -> UIActivityViewController {
        UIActivityViewController(activityItems: items, applicationActivities: nil)
    }

    func updateUIViewController(_ uiViewController: UIActivityViewController, context: Context) {}
}

File: android/app/src/main/java/com/cbtipul/app/data/PatientInvitation.kt
package com.cbtipul.app.data

import com.cbtipul.app.debug.InviteDebugLog
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.functions.functions
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.Headers
import io.ktor.http.HttpHeaders
import kotlinx.serialization.Serializable

@Serializable
enum class InvitationKind {
    @kotlinx.serialization.SerialName("initial") Initial,
}

@Serializable
data class PatientInvitation(
    val invitationId: String,
    val invitationUrl: String,
    val expiresAt: String,
)

@Serializable
enum class PatientInvitationPreviewStatus {
    @kotlinx.serialization.SerialName("valid") Valid,
    @kotlinx.serialization.SerialName("expired") Expired,
    @kotlinx.serialization.SerialName("claimed") Claimed,
    @kotlinx.serialization.SerialName("cancelled") Cancelled,
    @kotlinx.serialization.SerialName("invalid") Invalid,
}

@Serializable
data class PatientInvitationPreview(
    val status: PatientInvitationPreviewStatus,
    val therapistDisplayName: String? = null,
    val expiresAt: String? = null,
)

@Serializable
data class ClaimedPatientInvitation(
    val patientId: String,
    val therapistId: String,
)

@Serializable
private data class CreatePatientInvitationRequest(
    val patientId: String,
    val kind: InvitationKind = InvitationKind.Initial,
)

@Serializable
private data class GetPatientInvitationRequest(
    val token: String,
)

sealed class PatientInvitationClaimError : Exception() {
    data class Status(val status: PatientInvitationPreviewStatus) : PatientInvitationClaimError()
    data object Failed : PatientInvitationClaimError()
}

class PatientInvitationService(private val client: SupabaseClient) {
    suspend fun createPatientInvitation(patientId: String): PatientInvitation {
        if (!SupabaseConfig.isConfigured) throw IllegalStateException("not_configured")
        val http = client.functions.invoke(
            function = "create-patient-invitation",
            body = CreatePatientInvitationRequest(patientId = patientId),
            headers = jsonHeaders(),
        )
        return EdgePayload.json.decodeFromString(PatientInvitation.serializer(), http.bodyAsText())
    }

    suspend fun getPatientInvitation(token: String): PatientInvitationPreview {
        if (!SupabaseConfig.isConfigured) throw IllegalStateException("not_configured")
        val http = client.functions.invoke(
            function = "get-patient-invitation",
            body = GetPatientInvitationRequest(token = token),
            headers = jsonHeaders(),
        )
        return EdgePayload.json.decodeFromString(
            PatientInvitationPreview.serializer(),
            http.bodyAsText(),
        )
    }

    suspend fun claimPatientInvitation(token: String): ClaimedPatientInvitation {
        if (!SupabaseConfig.isConfigured) throw IllegalStateException("not_configured")
        try {
            val http = client.functions.invoke(
                function = "claim-patient-invitation",
                body = GetPatientInvitationRequest(token = token),
                headers = jsonHeaders(),
            )
            val claimed = EdgePayload.json.decodeFromString(
                ClaimedPatientInvitation.serializer(),
                http.bodyAsText(),
            )
            InviteDebugLog.d(
                "Claim response: patientId present = ${InviteDebugLog.present(claimed.patientId)}, " +
                    "therapistId present = ${InviteDebugLog.present(claimed.therapistId)}",
            )
            return claimed
        } catch (error: PatientInvitationClaimError) {
            throw error
        } catch (error: Exception) {
            InviteDebugLog.e("claim-patient-invitation", error)
            val body = EdgePayload.responseBody(error)
            val preview = runCatching {
                EdgePayload.json.decodeFromString(PatientInvitationPreview.serializer(), body)
            }.getOrNull()
            if (preview != null && preview.status != PatientInvitationPreviewStatus.Valid) {
                throw PatientInvitationClaimError.Status(preview.status)
            }
            throw PatientInvitationClaimError.Failed
        }
    }

    private fun jsonHeaders() = Headers.build {
        append(HttpHeaders.ContentType, ContentType.Application.Json.toString())
    }
}
